src/processor/operator/scan_node_id.cpp

```diff
@@ -46,10 +28,42 @@ pair<uint64_t, uint64_t> ScanNodeIDSharedState::getNextRangeToRead() {
     return make_pair(startOffset, startOffset + range);
 }
 
+void ScanNodeIDSharedState::initialize(Transaction* transaction) {
+    unique_lock lck{mtx};
+    if (initialized) {
+        return;
+    }
+    for (auto& tableState : tableStates) {
+        tableState->initialize(transaction);
+    }
+    initialized = true;
+}
+
+tuple<ScanTableNodeIDSharedState*, node_offset_t, node_offset_t>
+ScanNodeIDSharedState::getNextRangeToRead() {
+    unique_lock lck{mtx};
+    if (currentStateIdx == tableStates.size()) {
+        return make_tuple(nullptr, -1, -1);
+    }
+    auto [startOffset, endOffset] = tableStates[currentStateIdx]->getNextRangeToRead();
+    while (startOffset >= endOffset) {
+        currentStateIdx++;
+        if (currentStateIdx == tableStates.size()) {
+            return make_tuple(nullptr, -1, -1);
+        }
+        auto [_startOffset, _endOffset] = tableStates[currentStateIdx]->getNextRangeToRead();
+        startOffset = _startOffset;
+        endOffset = _endOffset;
+    }
+    auto tableState =
```

Comment:
I don't think you should check for `nullptr `here. Instead, you are not supposed to hit `currentStateIdx >= tableStates.size()`. I would put an assert here.